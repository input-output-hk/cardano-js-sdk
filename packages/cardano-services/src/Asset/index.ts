export * from './AssetHttpService';
export * from './CardanoTokenRegistry';
export * from './StubTokenMetadataService';
export * from './DbSyncNftMetadataService';
export * from './DbSyncAssetProvider';
export * from './TypeOrmNftMetadataService';
export * from './types';
