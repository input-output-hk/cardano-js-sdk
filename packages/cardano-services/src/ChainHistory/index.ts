export * from './ChainHistoryHttpService.js';
export * from './DbSyncChainHistory/index.js';
