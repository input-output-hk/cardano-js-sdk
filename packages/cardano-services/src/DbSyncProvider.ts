import { HealthCheckResponse, Provider } from '@cardano-sdk/core';
import { Pool } from 'pg';

const HEALTH_CHECK_QUERY = 'SELECT 1';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyArgs = any[];

export const DbSyncProvider = <
  T extends (abstract new (...args: AnyArgs) => {}) | (new (...args: AnyArgs) => {}) = { new (): {} }
>(
  BaseClass?: T
) => {
  abstract class Mixin extends (BaseClass || Object) implements Provider {
    public db: Pool;

    constructor(...args: AnyArgs) {
      const [db, ...baseArgs] = [...args];

      super(...baseArgs);

      this.db = db;
    }

    public async healthCheck(): Promise<HealthCheckResponse> {
      const result = await this.db.query(HEALTH_CHECK_QUERY);
      return { ok: !!result.rowCount };
    }
  }

  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
    ? A
    : T extends new (...baseArgs: infer A) => {}
    ? A
    : never;
  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
    ? I
    : T extends new (...baseArgs: AnyArgs) => infer I
    ? I
    : never;
  type ReturnedType = BaseInstance & { db: Pool; healthCheck: () => Promise<HealthCheckResponse> };

  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
    ? new (db: Pool, ...args: BaseArgs) => ReturnedType
    : abstract new (db: Pool, ...args: BaseArgs) => ReturnedType) & {
    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
  };
};

export type DbSyncProvider<T> = T extends new (...args: AnyArgs) => infer O ? O : never;
