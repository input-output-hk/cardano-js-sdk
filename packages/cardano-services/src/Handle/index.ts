export * from './HandleHttpService';
export * from './TypeOrmHandleProvider';
