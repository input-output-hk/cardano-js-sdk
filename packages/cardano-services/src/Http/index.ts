export * from './HttpServer';
export * from './HttpService';
export * from './types';
export * from './util';
