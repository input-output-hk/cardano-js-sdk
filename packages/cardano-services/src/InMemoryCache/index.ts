export * from './InMemoryCache';
export * from './NoCache';
export * from './defaults';
