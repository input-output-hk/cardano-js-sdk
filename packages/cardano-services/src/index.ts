export * from './Asset';
export * from './Http';
export * from './InMemoryCache';
export * from './Program';
export * from './RunnableModule';
export * from './StakePool';
export * from './TxSubmit';
export * from './Utxo';
export * from './ChainHistory';
export * from './Rewards';
export * as errors from './errors';
export * as util from './util';
