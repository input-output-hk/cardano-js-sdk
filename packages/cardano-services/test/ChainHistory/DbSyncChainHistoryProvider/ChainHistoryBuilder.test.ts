import { Cardano } from '@cardano-sdk/core';
import { ChainHistoryBuilder } from '../../../src';
import { Pool } from 'pg';
import { dummyLogger as logger } from 'ts-log';

describe('ChainHistoryBuilder', () => {
  let dbConnection: Pool;
  let builder: ChainHistoryBuilder;

  beforeAll(async () => {
    dbConnection = new Pool({ connectionString: process.env.POSTGRES_CONNECTION_STRING });
    builder = new ChainHistoryBuilder(dbConnection, logger);
  });

  afterAll(async () => {
    await dbConnection.end();
  });

  describe('queryTransactionInputsByHashes', () => {
    test('query transaction inputs by tx hashes', async () => {
      const result = await builder.queryTransactionInputsByHashes([
        Cardano.TransactionId('cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819'),
        Cardano.TransactionId('952dfa431223fd671c5e9e048e016f70fcebd9e41fcb726969415ff692736eeb')
      ]);
      expect(result).toHaveLength(2);
      expect(result).toMatchSnapshot();
    });
    test('query transaction inputs with empty array', async () => {
      const result = await builder.queryTransactionInputsByHashes([]);
      expect(result).toHaveLength(0);
    });
    test('query transaction inputs when tx hashes not found', async () => {
      const result = await builder.queryTransactionInputsByHashes([
        Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000')
      ]);
      expect(result).toHaveLength(0);
    });

    describe('with collateral true', () => {
      test('query transaction collateral inputs by tx hashes', async () => {
        const result = await builder.queryTransactionInputsByHashes(
          [
            Cardano.TransactionId('5acd6efb1b66299f1c5a2c4221af4bcaa4ba9929e8e6aa0e3f48707fa1796fc3'),
            Cardano.TransactionId('24e75c64a309fd8fb400933795b2522ca818cba80a3838c2ff14cec2cc8ffe4e')
          ],
          true
        );
        expect(result).toHaveLength(2);
        expect(result).toMatchSnapshot();
      });
      test('query transaction collateral inputs when tx not found or does not have any ', async () => {
        const result = await builder.queryTransactionInputsByHashes(
          [
            Cardano.TransactionId('295d5e0f7ee182426eaeda8c9f1c63502c72cdf4afd6e0ee0f209adf94a614e7'),
            Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000')
          ],
          true
        );
        expect(result).toHaveLength(0);
      });
    });
  });

  describe('queryTransactionOutputsByHashes', () => {
    test('query transaction outputs by tx hashes', async () => {
      const result = await builder.queryTransactionOutputsByHashes([
        Cardano.TransactionId('cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819'),
        Cardano.TransactionId('952dfa431223fd671c5e9e048e016f70fcebd9e41fcb726969415ff692736eeb')
      ]);
      expect(result).toHaveLength(4);
      expect(result).toMatchSnapshot();
    });
    test('query transaction outputs with empty array', async () => {
      const result = await builder.queryTransactionOutputsByHashes([]);
      expect(result).toHaveLength(0);
    });
    test('query transaction outputs when tx hashes not found', async () => {
      const result = await builder.queryTransactionOutputsByHashes([
        Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000')
      ]);
      expect(result).toHaveLength(0);
    });
  });

  describe('queryMultiAssetsByTxOut', () => {
    test('query multi assets by tx output', async () => {
      const result = await builder.queryMultiAssetsByTxOut([5_129_575n, 5_129_595n, 5_129_550n]);
      expect(result.size).toEqual(3);
      expect(result).toMatchSnapshot();
    });
    test('query multi assets with empty array', async () => {
      const result = await builder.queryMultiAssetsByTxOut([]);
      expect(result.size).toEqual(0);
    });
    test('query multi assets when tx output not found or has no multi-assets', async () => {
      const result = await builder.queryMultiAssetsByTxOut([0n, 5_396_852n]);
      expect(result.size).toEqual(0);
    });
  });

  describe('queryTxMintByHashes', () => {
    test('query transaction mint by tx hashes', async () => {
      const result = await builder.queryTxMintByHashes([
        Cardano.TransactionId('24e75c64a309fd8fb400933795b2522ca818cba80a3838c2ff14cec2cc8ffe4e'),
        Cardano.TransactionId('952dfa431223fd671c5e9e048e016f70fcebd9e41fcb726969415ff692736eeb')
      ]);
      expect(result.size).toEqual(2);
      expect(result).toMatchSnapshot();
    });
    test('query transaction mint with empty array', async () => {
      const result = await builder.queryTxMintByHashes([]);
      expect(result.size).toEqual(0);
    });
    test('query transaction mint when tx not found or has no mint operations', async () => {
      const result = await builder.queryTxMintByHashes([
        Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000'),
        Cardano.TransactionId('cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819')
      ]);
      expect(result.size).toEqual(0);
    });
  });

  describe('queryWithdrawalsByHashes', () => {
    test('query transaction withdrawals by tx hashes', async () => {
      const result = await builder.queryWithdrawalsByHashes([
        Cardano.TransactionId('cb66e0f5778718f8bfcfd043712f37d9993f4703b254a7a4d954d34225fe2f99')
      ]);
      expect(result.size).toEqual(1);
      expect(result).toMatchSnapshot();
    });
    test('query transaction withdrawals with empty array', async () => {
      const result = await builder.queryWithdrawalsByHashes([]);
      expect(result.size).toEqual(0);
    });
    test('query transaction withdrawals when tx not found or has no withdrawals', async () => {
      const result = await builder.queryWithdrawalsByHashes([
        Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000'),
        Cardano.TransactionId('cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819')
      ]);
      expect(result.size).toEqual(0);
    });
  });

  describe('queryRedeemersByHashes', () => {
    test('query transaction redeemers by tx hashes', async () => {
      const result = await builder.queryRedeemersByHashes([
        Cardano.TransactionId('5acd6efb1b66299f1c5a2c4221af4bcaa4ba9929e8e6aa0e3f48707fa1796fc3'),
        Cardano.TransactionId('24e75c64a309fd8fb400933795b2522ca818cba80a3838c2ff14cec2cc8ffe4e')
      ]);
      expect(result.size).toEqual(2);
      expect(result).toMatchSnapshot();
    });
    test('query transaction redeemers with empty array', async () => {
      const result = await builder.queryRedeemersByHashes([]);
      expect(result.size).toEqual(0);
    });
    test('query transaction redeemers when tx not found or has no redeemers', async () => {
      const result = await builder.queryRedeemersByHashes([
        Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000'),
        Cardano.TransactionId('cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819')
      ]);
      expect(result.size).toEqual(0);
    });
  });

  describe('queryCertificatesByHashes', () => {
    test('query certificates by tx hashes', async () => {
      const result = await builder.queryCertificatesByHashes([
        Cardano.TransactionId('face165bd7aa8d0d661cf1ceaa4e35d7611be3b1c7997da378c547aa2464a4fd'),
        Cardano.TransactionId('19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d'),
        Cardano.TransactionId('295d5e0f7ee182426eaeda8c9f1c63502c72cdf4afd6e0ee0f209adf94a614e7'),
        Cardano.TransactionId('68f21b8c6a8ae2c09839a00c908ba79bc81ffde240720f72589f842619915085'),
        Cardano.TransactionId('24986c0cd3475419bfb44756c27a0e13a6354a4071153f76a78f9b2d1e596089')
      ]);
      expect(result.size).toEqual(5);
      expect(result).toMatchSnapshot();
    });
    test('query certificates with empty array', async () => {
      const result = await builder.queryCertificatesByHashes([]);
      expect(result.size).toEqual(0);
    });
    test('query certificates when tx not found or has not certificates', async () => {
      const result = await builder.queryCertificatesByHashes([
        Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000'),
        Cardano.TransactionId('cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819')
      ]);
      expect(result.size).toEqual(0);
    });
  });
});
