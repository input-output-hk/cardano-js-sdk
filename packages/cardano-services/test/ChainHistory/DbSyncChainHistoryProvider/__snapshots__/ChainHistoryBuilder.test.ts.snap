// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChainHistoryBuilder queryCertificatesByHashes query certificates by tx hashes 1`] = `
Map {
  "face165bd7aa8d0d661cf1ceaa4e35d7611be3b1c7997da378c547aa2464a4fd" => Array [
    Object {
      "__typename": "PoolRetirementCertificate",
      "epoch": 179,
      "poolId": "pool1xv0p9vje3r09tac22mhgnsg08rmfm5qxadggjnfnq3cz52apdew",
    },
  ],
  "295d5e0f7ee182426eaeda8c9f1c63502c72cdf4afd6e0ee0f209adf94a614e7" => Array [
    Object {
      "__typename": "StakeKeyRegistrationCertificate",
      "stakeKeyHash": "c644b3906d4b9952886263832e59a841b5e1b073806747c8d731fbce",
    },
    Object {
      "__typename": "StakeKeyRegistrationCertificate",
      "stakeKeyHash": "8ebe40911f3ae2e6c1fe43dc647566ae1d8701d4d018a1d2a333124b",
    },
    Object {
      "__typename": "PoolRegistrationCertificate",
      "poolParameters": null,
    },
    Object {
      "__typename": "StakeDelegationCertificate",
      "poolId": "pool1y25deq9kldy9y9gfvrpw8zt05zsrfx84zjhugaxrx9ftvwdpua2",
      "stakeKeyHash": "c644b3906d4b9952886263832e59a841b5e1b073806747c8d731fbce",
    },
  ],
  "19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d" => Array [
    Object {
      "__typename": "PoolRegistrationCertificate",
      "poolParameters": null,
    },
    Object {
      "__typename": "StakeDelegationCertificate",
      "poolId": "pool1xv0p9vje3r09tac22mhgnsg08rmfm5qxadggjnfnq3cz52apdew",
      "stakeKeyHash": "34ff9be684ca26d2efee39a693fe051d393bb90d567357e6bbdeb2c8",
    },
  ],
  "24986c0cd3475419bfb44756c27a0e13a6354a4071153f76a78f9b2d1e596089" => Array [
    Object {
      "__typename": "MirCertificate",
      "pot": "treasury",
      "quantity": 100000000n,
      "rewardAccount": "stake_test1uzwd0ng8pw7vvhm4k3s28azx9c6ytug60uh35jvztgg03rge58jf8",
    },
    Object {
      "__typename": "MirCertificate",
      "pot": "treasury",
      "quantity": 100000000n,
      "rewardAccount": "stake_test1uq6p9hn9u53kvmh4mu98c0d4zzuekp2nkelnynct5g26lqs9yenqu",
    },
    Object {
      "__typename": "MirCertificate",
      "pot": "treasury",
      "quantity": 100000000n,
      "rewardAccount": "stake_test1upuzrnzp6qtwzl30dcm63sd63g724t6j639cpf2hknxymxqnt30fu",
    },
  ],
  "68f21b8c6a8ae2c09839a00c908ba79bc81ffde240720f72589f842619915085" => Array [
    Object {
      "__typename": "StakeKeyDeregistrationCertificate",
      "stakeKeyHash": "7c2fc08df5cb1e296b3b07d296731502296a381915869a047fb70c7b",
    },
  ],
}
`;

exports[`ChainHistoryBuilder queryMultiAssetsByTxOut query multi assets by tx output 1`] = `
Map {
  "5129550" => Map {
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652243484f43" => 10000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf39165224d494e54" => 5000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522524245525259" => 1000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 1828882901n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652256414e494c" => 15000000n,
  },
  "5129575" => Map {
    "22be35160908d953c25b30d1ad59876cbf073a8b31ef7a4f77468355414141" => 4208n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf39165224d494e54" => 1914111968n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522524245525259" => 1000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 1990000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652256414e494c" => 15000000n,
    "cd9d85fe5c8a1684c42ae0082e7e912fb7a7cca0900c5a3915eb1c3442454e59" => 2819431n,
    "d311d3488cc4fef19d05634adce8534977a3bc6fc18136ad65df1d4f6c71202e01" => 100000000n,
    "d311d3488cc4fef19d05634adce8534977a3bc6fc18136ad65df1d4f6c712038" => 172696733n,
  },
  "5129595" => Map {
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652243484f43" => 10000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf39165224d494e54" => 4865509476n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522524245525259" => 1000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 2000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652256414e494c" => 15000000n,
  },
}
`;

exports[`ChainHistoryBuilder queryRedeemersByHashes query transaction redeemers by tx hashes 1`] = `
Map {
  "5acd6efb1b66299f1c5a2c4221af4bcaa4ba9929e8e6aa0e3f48707fa1796fc3" => Array [
    Object {
      "executionUnits": Object {
        "memory": 4837876,
        "steps": 1615750737,
      },
      "index": 0,
      "purpose": "spend",
      "scriptHash": "9f244c64c6ca8e495f4d0ac48f106a1dbdf20b0821d598ebbcd1c57d",
    },
  ],
  "24e75c64a309fd8fb400933795b2522ca818cba80a3838c2ff14cec2cc8ffe4e" => Array [
    Object {
      "executionUnits": Object {
        "memory": 4195651,
        "steps": 1591910324,
      },
      "index": 0,
      "purpose": "spend",
      "scriptHash": "64de10aac4aad00a1a25079498378587c8d92825f17223dcf4fad59c",
    },
    Object {
      "executionUnits": Object {
        "memory": 88558,
        "steps": 46839629,
      },
      "index": 1,
      "purpose": "spend",
      "scriptHash": "6e4664e3f6acc48e849eacf4e3607ee6ae2758d6d8b80fa9d8fdd761",
    },
    Object {
      "executionUnits": Object {
        "memory": 88558,
        "steps": 46839629,
      },
      "index": 3,
      "purpose": "spend",
      "scriptHash": "6e4664e3f6acc48e849eacf4e3607ee6ae2758d6d8b80fa9d8fdd761",
    },
    Object {
      "executionUnits": Object {
        "memory": 88558,
        "steps": 46839629,
      },
      "index": 5,
      "purpose": "spend",
      "scriptHash": "6e4664e3f6acc48e849eacf4e3607ee6ae2758d6d8b80fa9d8fdd761",
    },
    Object {
      "executionUnits": Object {
        "memory": 106622,
        "steps": 55720739,
      },
      "index": 0,
      "purpose": "mint",
      "scriptHash": "22c3b86a5b88a78b5de52f4aed2831d1483b3b7681f1ee2569538130",
    },
  ],
}
`;

exports[`ChainHistoryBuilder queryTransactionInputsByHashes query transaction inputs by tx hashes 1`] = `
Array [
  Object {
    "address": "addr_test1qpcnmvyjmxmsm75f747u566gw7ewz4mesdw7yl278uf9r3f5l7d7dpx2ymfwlm3e56flupga8yamjr2kwdt7dw77ktyqqtx2r7",
    "id": "2791722",
    "index": 0,
    "txInputId": "19251f57476d7af2777252270413c01383d9503110a68b4fde1a239c119c4f5d",
    "txSourceId": "face165bd7aa8d0d661cf1ceaa4e35d7611be3b1c7997da378c547aa2464a4fd",
  },
]
`;

exports[`ChainHistoryBuilder queryTransactionInputsByHashes with collateral true query transaction collateral inputs by tx hashes 1`] = `Array []`;

exports[`ChainHistoryBuilder queryTransactionOutputsByHashes query transaction outputs by tx hashes 1`] = `
Array [
  Object {
    "address": "addr_test1qr74u5d3d07x53pq37acghulu7wvjt794s53xp2fz343drhx2aghqc4747n2yhu0dht5xz0t5j6n0fju3rs8mfj83tzqly49jr",
    "datum": undefined,
    "index": 0,
    "txId": "952dfa431223fd671c5e9e048e016f70fcebd9e41fcb726969415ff692736eeb",
    "value": Object {
      "assets": undefined,
      "coins": 5528822746696n,
    },
  },
  Object {
    "address": "addr_test1qzmpmxyl7prdf0pazt0futtrg6k9drcay3ednatxqm5ky854nqdumcu3xny9qu5sk9m0wm0ypyty84kxj0gwd9seppfqzs9z2n",
    "datum": undefined,
    "index": 1,
    "txId": "952dfa431223fd671c5e9e048e016f70fcebd9e41fcb726969415ff692736eeb",
    "value": Object {
      "assets": Map {
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652243484f43" => 10000000n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf39165224d494e54" => 5000000000n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522524245525259" => 1000000000n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 2000000000n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652256414e494c" => 15000000n,
      },
      "coins": 1000000000n,
    },
  },
  Object {
    "address": "addr_test1wrsexavz37208qda7mwwu4k7hcpg26cz0ce86f5e9kul3hqzlh22t",
    "datum": "c5dfa8c3cbd5a959829618a7b46e163078cb3f1b39f152514d0c3686d553529a",
    "index": 0,
    "txId": "cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819",
    "value": Object {
      "assets": Map {
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 10000000n,
      },
      "coins": 7420514n,
    },
  },
  Object {
    "address": "addr_test1qpvl4gf5r5h5w4yjrf58rt8slltem9tqfge36zqhsz5kgc8f2tc4ydn4zz9p0eymajqa4vj87he7yz6ssmh9x9h685hst9rgvk",
    "datum": undefined,
    "index": 1,
    "txId": "cefd2fcf657e5e5d6c35975f4e052f427819391b153ebb16ad8aa107ba5a3819",
    "value": Object {
      "assets": Map {
        "22be35160908d953c25b30d1ad59876cbf073a8b31ef7a4f77468355414141" => 4208n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf39165224d494e54" => 1914111968n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522524245525259" => 1000000000n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 1990000000n,
        "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652256414e494c" => 15000000n,
        "cd9d85fe5c8a1684c42ae0082e7e912fb7a7cca0900c5a3915eb1c3442454e59" => 2819431n,
        "d311d3488cc4fef19d05634adce8534977a3bc6fc18136ad65df1d4f6c71202e01" => 100000000n,
        "d311d3488cc4fef19d05634adce8534977a3bc6fc18136ad65df1d4f6c712038" => 172696733n,
      },
      "coins": 850065083n,
    },
  },
]
`;

exports[`ChainHistoryBuilder queryTxMintByHashes query transaction mint by tx hashes 1`] = `
Map {
  "952dfa431223fd671c5e9e048e016f70fcebd9e41fcb726969415ff692736eeb" => Map {
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652243484f43" => 10000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf39165224d494e54" => 5000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522524245525259" => 1000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf3916522534245525259" => 2000000000n,
    "57fca08abbaddee36da742a839f7d83a7e1d2419f1507fcbf391652256414e494c" => 15000000n,
  },
  "24e75c64a309fd8fb400933795b2522ca818cba80a3838c2ff14cec2cc8ffe4e" => Map {
    "22c3b86a5b88a78b5de52f4aed2831d1483b3b7681f1ee2569538130f32324c10a93fe1962cf7470543932cc064482c3495fe3de7502f9edd84ca114" => 15355454n,
  },
}
`;

exports[`ChainHistoryBuilder queryWithdrawalsByHashes query transaction withdrawals by tx hashes 1`] = `
Map {
  "cb66e0f5778718f8bfcfd043712f37d9993f4703b254a7a4d954d34225fe2f99" => Array [
    Object {
      "quantity": 1409612n,
      "stakeAddress": "stake_test1uqvszkuv7gwzcfshrm37ywcxuqhh7mqx95ey3cm3cx8s76c7w97f8",
    },
    Object {
      "quantity": 229082n,
      "stakeAddress": "stake_test1urpklgzqsh9yqz8pkyuxcw9dlszpe5flnxjtl55epla6ftqktdyfz",
    },
  ],
}
`;
